### PR TITLE
[BUGFIX] Fix for docker(-compose) to allow the app to be run as a regular user.

### LINKED ISSUE #857: [Help]: Change docker container user chown maintenance issue
### 📝 Describe your problem

Currently `docker exec uptime whoami` results in `root`. I want to run this whole container as a different user using this dockerfile.

````
FROM louislam/uptime-kuma:1
RUN adduser --system --group uptime
USER uptime
````

But I get a bunch of:
```
==> Performing startup jobs and maintenance tasks
chown: changing ownership of '/app/data/kuma.db': Operation not permitted
chown: changing ownership of '/app/data/kuma.db-wal.bak20211022112928': Operation not permitted
chown: changing ownership of '/app/data/kuma.db-wal.bak20211101090108': Operation not permitted
chown: changing ownership of '/app/data/kuma.db-shm.bak20211101090108': Operation not permitted
chown: changing ownership of '/app/data/kuma.db.bak20211022112928': Operation not permitted
chown: changing ownership of '/app/data/kuma.db.bak20211101090108': Operation not permitted
chown: changing ownership of '/app/data/kuma.db.bak0': Operation not permitted
chown: changing ownership of '/app/data/kuma.db-wal.bak0': Operation not permitted
chown: changing ownership of '/app/data/kuma.db-shm.bak20211022112928': Operation not permitted
chown: changing ownership of '/app/data/upload': Operation not permitted
chown: changing ownership of '/app/data/error.log': Operation not permitted
chown: changing ownership of '/app/data/kuma.db-shm.bak0': Operation not permitted
chown: changing ownership of '/app/data': Operation not permitted
``` 

This is because the user is switched and then apparently a miantenance task starts doing chown commands. How do I disable this task? Because I want to run it before my `USER` line instead. Something like 
````
RUN adduser --system --group uptime && \
    mkdir -p /app/data && \
    chown -R uptime:uptime /app
USER uptime
````

### 🐻 Uptime-Kuma version

1

### 💻 Operating System

Ubuntu 20

### 🌐 Browser

Chrome

### 🐋 Docker

_No response_

### 🏷️ Docker Image Tag

louislam/uptime-kuma:1

### 🟩 NodeJS Version

_No response_

### ⚠️ Please verify that this question has NOT been raised before.

- [X] I checked and didn't find similar question

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### PR DESCRIPTION
**Pull request by justSem**
# Description

Fixes #857 and #694 

Due to the calling of `setpriv` as an regular user and some weird env variable stuff which seemed set wrongly to me the application couldn't be launched in docker as a non-root user.
(And really, an application like this _shouldn't_ run as root as it simply doesn't need it)

The proposed fix properly sets the environment variables reflecting the actual UID and GID and asseses this before launching.

>An user switching an existing setup from root to non-root user might need to manually `chown` the `/app/data` directory once depending on how they set up the storage. This does __not__ involve altering any container files.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I ran ESLint and other linters for modified files
  - Not applicable. I modified a bash file ;)
- [X] I have performed a self-review of my own code and tested it
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings

## Screenshots (if any)

Please do not use any external image service. Instead, just paste in or drag and drop the image here, and it will be uploaded automatically.

### CODE REVIEW

**Comment by k3rnelpan1c-dev** [comment]
While this is certainly one way to handle rootles container or containers that run as non root account due to CLI params (i.e. `--user 1001:1001`), would it not be a more standard way to let the container run as a non root user by default?

What I mean by that is that most of the entrypoint script could be made obsolete and to simply create a user+group during the container build and ensure that the `/data` dir gets creted and properly assigned.

> As I have to deal with this a lot at work some reference from a platform that enforces the container to run as non root unless you explicitly allow it otherwise: [OpenShift has a section in their docs on how to build a container so it supports arbitrary user ids](https://docs.openshift.com/container-platform/4.9/openshift_images/create-images.html#use-uid_create-images)

I know this works, since I ended up building a custom image that uses the recomendation from the docs as well as what Nginx uses for their unpribileged container to test `uptime-kuma` in an unprivileged enforcing environment.
https://github.com/nginxinc/docker-nginx-unprivileged/blob/main/stable/debian/Dockerfile#L20-L21

However, the question remains if this would be the 'better' way or to keep the entrypoint script and use the fix proposed by this PR.

Either way, these are just my two cents to the matter that came up when I saw that you where faster than me to propose changes :slightly_smiling_face:

**Comment by k3rnelpan1c-dev** [comment]
from what I can tell the scrip has nothing NodeJS specific.
It really just ensures that the `/data` dir exists and has its owner set 'properly' (with your fix it actually does that, b4 you would have had to specify `PUID` and `PUID`) and spawns of the NodeJS process as the same user the data dir now has as its owner.

At least that is what I could gather from it :)

Regardless, all I wanted to do is throw in the comment and open a discussion which way is the preferable one.

Either works since even if you where to create a user within the container build and run by default as that a user could still use the container runtime CLI (Docker, Podman, ...) to overwrite the user and group the container will run as would then lead to the NodeJS process to run as those ids.

There is only one instance where I am not 100% certain how it would react, if a user where to use a docker volume to mount on a pre-created `/data` dir (created at container build time and owned by the user created at build time) but setting a different UID and GID via the CLI or a thing such as Compose.
The likely 'solution' for that would be to set the `/data` dir at creation to 777.
(hope that made sense)

**Comment by k3rnelpan1c-dev** [comment]
Thanks for testing my assumed edge case :sweat_smile: 
And yeah I assumed as much that it would run into this issue.
Setting the dir to 777 is NEVER a good idea so my apologies for bribing it up at all, was just throwing it out there for the only way to avoid permission errors I suppose.

I like the idea of giving the entrypoint script the porous of a environment checker and presenting the directions to fix possible issues (i.e. legacy permissions or plain wrong bind mount perms).

**Comment by justSem** [comment]
> Thanks for testing my assumed edge case 😅 And yeah I assumed as much that it would run into this issue. Setting the dir to 777 is NEVER a good idea so my apologies for bringing it up at all, was just throwing it out there for the only way to avoid permission errors I suppose.
> 
> I like the idea of giving the entrypoint script the porous of a environment checker and presenting the directions to fix possible issues (i.e. legacy permissions or plain wrong bind mount perms).

I've altered the script to (attempt) to work around this edge-case.
It now assesses ownership, directory and file permissions, and tries to sets them properly.

If it fails it'll exit with an error message instructing on how to resolve the problem.

File: extra/entrypoint.sh [hunk 1]
```diff
@@ -2,20 +2,54 @@
...
+        echo "File ownership incorrect, attempting to fix."
+        chown -hRc "$(id -u)":"$(id -g)" /app/data || echo "ERROR: Failed to set file ownership. Please run 'sudo chown -R $(id -u):$(id -g) /path/to/container/volume' to resolve."; exit 1
+    fi
+
+   # Checks for R/W permissions
+   if [ $(stat -c%a /app/data) -ne 770 ]; then
+        echo "Directory permissions incorrect, attempting to fix."
+        find /app/data -type d -exec chmod 770 {} \;
+
+        #Re-run the check
```
Comment (Saibamen) [inline extra/entrypoint.sh:19]: Be consistent in comments
```suggestion
        # Re-run the check
```

File: extra/entrypoint.sh [hunk 2]
```diff
@@ -2,20 +2,54 @@
...
-PGID=${PGID=0}
+
 
 files_ownership () {
-    # -h Changes the ownership of an encountered symbolic link and not that of the file or directory pointed to by the symbolic link.
-    # -R Recursively descends the specified directories
-    # -c Like verbose but report only when a change is made
-    chown -hRc "$PUID":"$PGID" /app/data
+    # Check if the /app/data folder is owned by the user invoking the container
+    if [ $(stat -c%u /app/data) !=  $(id -u) ]; then
```
Comment (Saibamen) [inline extra/entrypoint.sh:9]: It would be great to have `/app/data` in variable (used 9 times here)

File: extra/entrypoint.sh [hunk 3]
```diff
@@ -2,20 +2,54 @@
...
+
+        #Re-run the check
+        if [ $(stat -c%a /app/data) -ne 770 ]; then
+                echo "ERROR: Failed to set file permissions. Please run 'sudo find /path/to/container/volume -type d chmod 770 {} \;' to resolve."
+                exit 1
+        fi
+        echo "Directory permission fix succesful! Continuing."
+   fi
+
+   #Check the R/W permissions on the files
```
Comment (Saibamen) [inline extra/entrypoint.sh:27]: Be consistent in comments
```suggestion
   # Check the R/W permissions on the files
```

File: extra/entrypoint.sh [hunk 4]
```diff
@@ -2,20 +2,54 @@
...
+   if [ $(stat -c%a /app/data/* | head -n 1) != 640 ]; then
+        echo "File permissions incorrect. Attempting to fix."
+        find /app/data -type f -exec chmod 640 {} \;
+
+        #Re-run the check
```
Comment (Saibamen) [inline extra/entrypoint.sh:32]: Be consistent in comments
```suggestion
        # Re-run the check
```

File: extra/entrypoint.sh [hunk 5]
```diff
@@ -2,20 +2,54 @@
...
 files_ownership
 
-echo "==> Starting application with user $PUID group $PGID"
+echo "==> Starting application as user: $(id -u) ($USER) and group $(id -g)"
 
 # --clear-groups Clear supplementary groups.
-exec setpriv --reuid "$PUID" --regid "$PGID" --clear-groups "$@"
+if [ $(id -u) -eq 0 ];
+then
+        #We're running as root, so we can use setpriv without problems.
```
Comment (Saibamen) [inline extra/entrypoint.sh:50]:
```suggestion
        # We're running as root, so we can use setpriv without problems.
```

File: extra/entrypoint.sh [hunk 6]
```diff
@@ -2,20 +2,54 @@
...
+        exec setpriv --reuid "$PUID" --regid "$PGID" --clear-groups "$@"
+else
+        #We're running as a regular user, so we'll launch the app as one.
```
Comment (Saibamen) [inline extra/entrypoint.sh:53]:
```suggestion
        # We're running as a regular user, so we'll launch the app as one.
```

File: extra/entrypoint.sh [hunk 7]
```diff
@@ -2,20 +2,54 @@
...
 files_ownership () {
-    # -h Changes the ownership of an encountered symbolic link and not that of the file or directory pointed to by the symbolic link.
-    # -R Recursively descends the specified directories
-    # -c Like verbose but report only when a change is made
-    chown -hRc "$PUID":"$PGID" /app/data
+    # Check if the /app/data folder is owned by the user invoking the container
+    if [ $(stat -c%u /app/data) !=  $(id -u) ]; then
+        echo "File ownership incorrect, attempting to fix."
+        chown -hRc "$(id -u)":"$(id -g)" /app/data || echo "ERROR: Failed to set file ownership. Please run 'sudo chown -R $(id -u):$(id -g) /path/to/container/volume' to resolve."; exit 1
+    fi
```
Comment (Saibamen) [inline extra/entrypoint.sh:12]: We are missing echo for "success" here. Directory and file permission has this message
```suggestion
        chown -hRc "$(id -u)":"$(id -g)" /app/data || echo "ERROR: Failed to set file ownership. Please run 'sudo chown -R $(id -u):$(id -g) /path/to/container/volume' to resolve."; exit 1
        echo "File ownership fix succesful! Continuing."
    fi
```

**Review by Saibamen — CHANGES_REQUESTED** [review]

**Review by Saibamen — APPROVED** [review]

**Comment by justSem** [comment]
@Saibamen Thank you for your feedback and approval! :)

**Comment by gaby** [comment]
@justSem Doesn't running as a regular user break the ability of sending pings? I remember a long time ago someone else tried this and it broke the Docker image.

**Comment by justSem** [comment]
> @justSem Doesn't running as a regular user break the ability of sending pings? I remember a long time ago someone else tried this and it broke the Docker image. 

No, ping isn't a privileged unix process.
I run my instance with this script and ping works just fine

**Comment by k3rnelpan1c-dev** [comment]
While I don't run this script yet, I have Uptime-Kuma deployed in the `restrictive` SCC (Security Context Constraint) of Open Shift and it works just fine.

*I built my own image with a slightly altered version of this script though, since the `restricted` SCC wont allow for any privileged or root running pod, even if its just at startup.

**Comment by gaby** [comment]
@justSem Here https://github.com/louislam/uptime-kuma/issues/488 it was a long time ago though.

**Comment by Kurt108** [comment]
@k3rnelpan1c-dev  could you share your script? Getting Kuma running on Openshift and with a persistent Volume is now as easy as it should be :-o

**Comment by openstep** [comment]
> While I don't run this script yet, I have Uptime-Kuma deployed in the `restrictive` SCC (Security Context Constraint) of Open Shift and it works just fine.
> 
> *I built my own image with a slightly altered version of this script though, since the `restricted` SCC wont allow for any privileged or root running pod, even if its just at startup.

Hi,

could you please share the info necessary running this in OpenShift? 
Many thanks.

**Comment by k3rnelpan1c-dev** [comment]
For those curious, yes I will share my config for OCP in a new issue, since it is just marginal connected to this PR.
Sorry to have disrupted the PR topic by my comment :sweat_smile:

**Comment by beneiltis** [comment]
When are you gonne merge this? :-)

**Comment by kaysond** [comment]
This would solve my problem of not having root access to my fs!

**Comment by justSem** [comment]
@Saibamen @louislam Is this going to get merged on any point? Or are there any changes required?

**Comment by Saibamen** [comment]
@justSem: I don't have write permissions

**Comment by OlivierAlves** [comment]
May someone merge this it would be really helpful... Thanks !

**Comment by SirSirae** [comment]
@louislam may someone merge this it would be really helpful ! Thank you !

**Comment by kaysond** [comment]
Bump. @louislam can we please get this into the next release?

**Comment by kaysond** [comment]
Bumping again and pinging @louislam

**Comment by Marck** [comment]
@louislam any change this could come in a release?
I would love to see, I have a volume that is mounted under a non-root user for all my containers use but I cannot use that because this runs as root (which is not needed)

**Review by ScmTble — APPROVED** [review]

**Comment by chakflying** [comment]
Obsolete by #4052.

**Pull request closed without merging**